### PR TITLE
Refactor number of pages alias (fix #1090)

### LINKED ISSUE #1090: {nb} breaks if text shaping is turned on with certain fonts
The special `{nb}` code fails with some fonts when text shaping is turned on.

**Minimal code**
Please include some minimal Python code reproducing your issue:
```python
pdf = FPDF(format='letter')
pdf.add_font('gentium', style='', fname='GenBkBasR.ttf')
pdf.add_page()
pdf.set_font('gentium', '', 24)
pdf.set_text_shaping(True)
pdf.write(text='Pages {nb}')
pdf.ln()
pdf.set_text_shaping(False)
pdf.write(text='Pages {nb}')
pdf.output('test.pdf')
```
**Result**
image
**Environment**
Please provide the following information:
* **Operating System**: Ubuntu
* **Python version**: 11
* **`fpdf2` version used**: 2.7.7

### PR DESCRIPTION
**Pull request by andersonhc**
e.g. Fixes #0 

**Checklist**:

- [x] The GitHub pipeline is OK (green), 
      meaning that both `pylint` (static code analyzer) and `black` (code formatter) are happy with the changes of this PR.

- [x] A unit test is covering the code added / modified by this PR

- [x] This PR is ready to be merged 

- [x] A mention of the change is present in `CHANGELOG.md`

By submitting this pull request, I confirm that my contribution is made under the terms of the [GNU LGPL 3.0 license](https://github.com/py-pdf/fpdf2/blob/master/LICENSE).

### CODE REVIEW

File: fpdf/fpdf.py [hunk 1]
```diff
@@ -279,7 +279,9 @@ def __init__(
         but is less compatible with the PDF spec.
         """
         self.page = 0  # current page number
-        self.pages = {}  # array of PDFPage objects starting at index 1
+        self.pages: Dict[int, PDFPage] = (
+            {}
+        )  # array of PDFPage objects starting at index 1
```
Comment (Lucas-C) [inline fpdf/fpdf.py:284]:
```suggestion
        # array of PDFPage objects starting at index 1:
        self.pages: Dict[int, PDFPage] = {}
```
Just a suggestion to make the code clearer

File: fpdf/line_break.py [hunk 2]
```diff
@@ -332,6 +333,25 @@ def render_pdf_text_core(self, frag_ws, current_ws):
...
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.uuid = uuid4()
+
+    def get_alias_string(self):
+        return f"::alias:{self.uuid}::"
+
+    def render_pdf_text(self, *args, **kwargs):
+        self._render_args = args
+        self._render_kwargs = kwargs
```
Comment (Lucas-C) [inline fpdf/line_break.py:347]: Shouldn't those attributes be defined in the constructor?

File: fpdf/fpdf.py [hunk 3]
```diff
@@ -5158,8 +5175,14 @@ def output(
             # Generating .buffer based on .pages:
             if self._toc_placeholder:
                 self._insert_table_of_contents()
-            if self.str_alias_nb_pages:
```
Comment (Lucas-C) [inline fpdf/fpdf.py:5161]: Why removing this IF?

**Review by Lucas-C — APPROVED** [review]

**Review by gmischler — APPROVED** [review]
Alias = "A different **_name_** for something"

That's not really what we're dealing with here, at all.

Yes, I know that `.str_alias_nb_pages` has had that name for like forever, but maybe we shouldn't propagate it to all the internal code dealing with text substitution.
In #1202 I tried to use terminology that describes better what really happens. I probably haven't found the absolute best wording, but it may help as a more reasonable starting point.

Using names that don't actually reflect the purpose of the named elements makes the code really confusing to read and hard to understand. I'm not a native english speaker, but I like to be as precise as possible about those things.

Other than that, I expect the result to "just work"... 😉

File: fpdf/line_break.py (hunk 2 above)
Comment (andersonhc) [reply fpdf/line_break.py:393]: I added docstrings to explain this method

**Comment by andersonhc** [comment]
@Lucas-C @gmischler 
This PR should be ready now.
I renamed the internal methods from "alias" to "text substitution" but didn't touch the ones that already existed (like alias_nb_pages) to keep backwards compatibility.

Lucas, I changed the text shaping enable/disable test you created recently because after this change the replaced number of pages is similar with text shaping on or off. I replaced by a text with ligatures that have internal and visual differences.

**Pull request merged**